zerver/actions/users.py

```diff
@@ -82,98 +87,21 @@
 
 
 def do_delete_user(user_profile: UserProfile, *, acting_user: UserProfile | None) -> None:
-    if user_profile.realm.is_zephyr_mirror_realm:
-        raise AssertionError("Deleting zephyr mirror users is not supported")
-
-    do_deactivate_user(user_profile, acting_user=acting_user)
-
-    to_resubscribe_recipient_ids = set(
-        Subscription.objects.filter(
-            user_profile=user_profile, recipient__type=Recipient.DIRECT_MESSAGE_GROUP
-        ).values_list("recipient_id", flat=True)
-    )
-    user_id = user_profile.id
-    realm = user_profile.realm
-    date_joined = user_profile.date_joined
-    personal_recipient = user_profile.recipient
-
-    with transaction.atomic(durable=True):
-        user_profile.delete()
-        # Recipient objects don't get deleted through CASCADE, so we need to handle
-        # the user's personal recipient manually. This will also delete all Messages pointing
-        # to this recipient (all direct messages sent to the user).
-        assert personal_recipient is not None
-        personal_recipient.delete()
-        replacement_user = create_user(
-            force_id=user_id,
-            email=Address(
-                username=f"deleteduser{user_id}", domain=get_fake_email_domain(realm.host)
-            ).addr_spec,
-            password=None,
-            realm=realm,
-            full_name=f"Deleted User {user_id}",
-            active=False,
-            is_mirror_dummy=True,
-            force_date_joined=date_joined,
-        )
-        subs_to_recreate = [
-            Subscription(
-                user_profile=replacement_user,
-                recipient=recipient,
-                is_user_active=replacement_user.is_active,
-            )
-            for recipient in Recipient.objects.filter(id__in=to_resubscribe_recipient_ids)
-        ]
-        Subscription.objects.bulk_create(subs_to_recreate)
+    do_delete_user_core(user_profile, delete_messages=True, acting_user=acting_user)
 
-        RealmAuditLog.objects.create(
-            realm=replacement_user.realm,
-            modified_user=replacement_user,
-            acting_user=acting_user,
-            event_type=AuditLogEventType.USER_DELETED,
-            event_time=timezone_now(),
-        )
 
+def do_delete_user_preserving_messages(
+    user_profile: UserProfile, *, acting_user: UserProfile | None
+) -> None:
+    do_delete_user_core(user_profile, delete_messages=False, acting_user=acting_user)
 
-def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
```

Comment:
The displayed diff is pretty much unreadable like this - for review, it's better to just checkout the new code directly and review the `do_delete_user_core` function.